debezium-server-voyager/debezium-server-voyagerexporter/src/main/java/io/debezium/server/ybexporter/EventDedupCache.java

```diff
@@ -0,0 +1,162 @@
+
+public class EventDedupCache {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventQueue.class);
+    private static final String EOF_MARKER = "\\.";
+    String dataDir;
+    private LinkedList<String> mostRecentIdFirstList = new LinkedList<>();
+    private Set<Integer> cache = new HashSet<>();
+    private long currentQueueSegmentIndex = 0;
+    private static final String QUEUE_SEGMENT_FILE_NAME = "segment";
+    private static final String QUEUE_SEGMENT_FILE_EXTENSION = "ndjson";
+    private static final String QUEUE_FILE_DIR = "queue";
+    private long maxCacheSize = 1000000;
+    private String currentQueueSegmentPath;
+
+    public EventDedupCache(String dataDir) {
+        this.dataDir = dataDir;
+    }
+
+    public void warmUp() {
+        cache = new HashSet<>();
+        recoverLatestQueueSegment();
+        if (currentQueueSegmentPath == null) {
+            LOGGER.info("No queue segment found. Nothing to recover into event cache");
+            return;
+        }
+        while (true) {
+            try (ReverseFileReader reverseReader = new ReverseFileReader(currentQueueSegmentPath)) {
+                while (reverseReader.hasNext()) {
+                    String event = reverseReader.nextLine();
+                    if (event == null) {
+                        // Segment is empty
+                        break;
+                    }
+                    if (event.equals(EOF_MARKER) || event.isEmpty() || event.isBlank()) {
+                        continue;
+                    }
+                    String cacheMetadata = getCacheMetadata(event);
+                    if (cacheMetadata == null) {
+                        continue;
+                    }
+                    cache.add(cacheMetadata.hashCode());
+                    mostRecentIdFirstList.addLast(cacheMetadata);
+                    if (cache.size() >= maxCacheSize) {
+                        break;
+                    }
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            if (cache.size() >= maxCacheSize) {
+                break;
+            }
+
+            // Segment is exhausted. Move to previous segment
+            currentQueueSegmentIndex--;
+            if (currentQueueSegmentIndex < 0) {
+                // No more segments to read
+                break;
+            }
+            currentQueueSegmentPath = getFilePathWithIndex(currentQueueSegmentIndex);
+        }
+        LOGGER.info("Recovered {} events into event cache", cache.size());
+    }
+
+    private String getFilePathWithIndex(long index) {
+        String queueSegmentFileName = String.format("%s.%d.%s", QUEUE_SEGMENT_FILE_NAME, index,
+                QUEUE_SEGMENT_FILE_EXTENSION);
+        return Path.of(dataDir, QUEUE_FILE_DIR, queueSegmentFileName).toString();
+    }
+
+    // Duplicate in EventQueue.java
+    private void recoverLatestQueueSegment() {
```

Comment:
Hmm, just thinking about the class relationships. How about the following:

1. `EventDedupCache` is part of the `EventQueue` . It is not used by any other class (the deduplication logic is only relevant for the `EventQueue`.) Before writing the event to file, the `EventQueue` checks with the `EventDedupCache` and after writing, it informs the `EventDedupCache`. 
2. As part of `EventQueue`'s initialization process, it initializes `EventDedupCache`, and passes it as many historical `EventQueueSegments` as the `EventDedupCache` would need. If it is creating a cache that can hold 1Million events, it figures out how many queue segments to pass to the init of `EventDedupCache`. `EventDedupCache` reads from those segments and warms up it's cache. (you can figure the no. of events in each segment from metadb)
3. Also, I think you can drop the "cache" word in the name `EventDedupCache`. The function of the class is to dedup events, so you can call it `EventDedup`; whether it uses a cache or something else is an internal detail of the class.

Reply:
The usage of "cache" terminology in this context is common:
- https://www.ibm.com/docs/en/spectrum-protect/8.1.9?topic=reference-dedupcachesize
- https://vector.dev/docs/reference/configuration/transforms/dedupe/#cache.num_events
- https://utcc.utoronto.ca/~cks/space/blog/unix/NFSServerReplyCacheWhy

Reply:
Let's go with the class organisation that Aneesh is suggesting here. That will remove a lot of code duplication and even simplify it to a great extent.
The `EventQueue` should use metadb to figure out the ordered list of `EventQueueSegments` and pass it to the `EventDedupeCache`.